main.py

```diff
@@ -1,74 +1,77 @@
-				canvas.SetPixel(x + i, y + j, 255, 255, 255)
-			elif j == 7 and (i == 5 or i == 6):
-				canvas.SetPixel(x + i, y + j, 255, 255, 255)
-			else:
-				canvas.SetPixel(x + i, y + j, 117, 119, 122)
-
-def run():
-	logging.info("Starting l-train-display.")
-
-	options = RGBMatrixOptions()
-	options.rows = 32
-	options.cols = 64
-	options.chain_length = 1
-	options.parallel = 1
-	options.hardware_mapping = "adafruit-hat"
-
-	matrix = RGBMatrix(options=options)
-
-	dir_font = graphics.Font()
-	dir_font.LoadFont("5x8.bdf")
-	time_font = graphics.Font()
-	time_font.LoadFont("5x8.bdf")
-
-	l_train_color = graphics.Color(167, 169, 172)
-	font_color = graphics.Color(250, 200, 50)
-
-	canvas = matrix.CreateFrameCanvas()
-
-	while True:
-		manhattan_in, brooklyn_in = get_next_trains()
-		canvas.Clear()
-
-		draw_l_train_logo(2, 3, canvas)
-		draw_l_train_logo(2, 18, canvas)
-		graphics.DrawText(canvas, dir_font, 14, 12, font_color, "MANH")
-		graphics.DrawText(canvas, dir_font, 14, 27, font_color, "BKLN")
-
-		manhattan_offset = 37 if len(manhattan_in) == 5 else 42
-		brooklyn_offset = 37 if len(brooklyn_in) == 5 else 42
-
-		graphics.DrawText(canvas, time_font, manhattan_offset, 12, font_color, manhattan_in)
-		graphics.DrawText(canvas, time_font, brooklyn_offset, 27, font_color, brooklyn_in)
-		canvas = matrix.SwapOnVSync(canvas)
-
-		time.sleep(30)
+import time, sys, threading
+from train_updater import TrainUpdater, Trip, Direction
+from train_displayer import TrainDisplayer
+import mta_api_key
+import constants
+
+updater = TrainUpdater(constants.MTA_API_URL, mta_api_key.key)
+displayer = TrainDisplayer("5x8.bdf")
+
+lock = threading.Lock()
+next_trains_trip = {
+    "North": None,
+    "South": None
+}
+
+def parse_arrival_time(arrival):
+    now = time.time()
+    arrival_in_min = round((arrival - now)/60)
+    if arrival_in_min < -5:
+        return "-"
+    else:
+        return str(arrival_in_min)
+
+def parse_trips_and_update_display():
+    while True:
+        lock.acquire()
+        trip_north = next_trains_trip["North"]
+        trip_south = next_trains_trip["South"]
+        lock.release()
+
+        north_stop = "No Trains"
+        south_stop = "No Trains"
+        north_time = "-"
+        south_time = "-"
+        if trip_north is not None:
+            north_stop = constants.L_STOPS[trip_north.terminus[:-1]]
+            north_time = parse_arrival_time(trip_north.next_train)
+        if trip_south is not None:
+            south_stop = constants.L_STOPS[trip_south.terminus[:-1]]
+            south_time = parse_arrival_time(trip_south.next_train)
+
+        displayer.update_display(north_stop, south_stop, north_time, south_time)
+        time.sleep(constants.DISPLAY_SCROLL_SPEED)
+
+def update_next_trains():
+    while True:
+        train_north, train_south = updater.get_next_trains()
+        lock.acquire()
+        next_trains_trip["North"] = train_north
+        next_trains_trip["South"] = train_south
+        lock.release()
```

Comment:
It may be safer to use the lock from a with statement instead of manually acquiring and releasing

```suggestion
        with lock:
            next_trains_trip["North"] = train_north
            next_trains_trip["South"] = train_south
```